llvm/lib/SYCLLowerIR/SpecConstants.cpp

```diff
@@ -662,44 +624,89 @@ Instruction *emitSpecConstantRecursiveImpl(Type *Ty, Instruction *InsertBefore,
     Elements.push_back(Def);
     Index++;
   };
-  auto LoopIteration = [&](Type *ElTy, unsigned LocalIndex) {
-    const auto ElemDefaultValue = getElemDefaultValue(
-        Ty, ElTy, DefaultValue, LocalIndex, M.getDataLayout());
-
+  auto LoopIteration = [&](Type *ElTy, unsigned LocalOffset) {
+    auto ElOffset = CurrentOffset + LocalOffset;
+    auto It = llvm::lower_bound(DefinedElements, ElOffset,
+                                [](const std::pair<uint64_t, Constant *> &LHS,
+                                   uint64_t RHS) { return LHS.first < RHS; });
     // If the default value is a composite and has the value 'undef', we should
     // not generate a bunch of __spirv_SpecConstant for its elements but
     // pass it into __spirv_SpecConstantComposite as is.
-    if (isa<UndefValue>(ElemDefaultValue))
-      HandleUndef(ElemDefaultValue);
+    if (It == DefinedElements.end() || It->first != ElOffset)
+      HandleUndef(UndefValue::get(ElTy));
```

Comment:
Will be changed to `Constant::getNullValue(ElTy)` in https://github.com/intel/llvm/pull/17341